src/ztex_md5crypt.c

```diff
@@ -0,0 +1,249 @@
+	/*
+	{"$1$fkEasaUI$G7CelOWHkol2nVHN8XQP40", "aaaaaaaaaaaaaaaaaaaaaaaaa"},
+	{"$1$12345678$/Y8HRXkjaI0wpCjIOG1xv1", "key_len=31..................../"},
+	{"$1$12345678$YwDJjafnp9d0vGAuGWwaK/", "key_len=32...................../"},
+	{"$1$IsuapfCX$4Yq0Adq5nNZgl0LwbSl5Y0", "aaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaa"},
+	{"$1$rSZfNcKX$N4XPvGrfhKsyoEcRSaqmG0", "aaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaa"},
+	{"$1$abcd0000$nNTpR9XboutDDILjNSXMC/",
+		"key_len=64.....567890123456789012345678901234567890123456789abcd"},
+	*/
+	{NULL}
+};
+
+
+static void init(struct fmt_main *fmt_main)
+{
+//printf("verbosity (init):%d\n", options.verbosity);
+
+	device_format_init(fmt_main, &bitstream, options.acc_devices);//,
+		//options.verbosity);
+}
+
+
+/*
+ * There's an existing CPU implementation. It stores salt in the database.
+ * 1) Salt data structure db_salt->salt is specific to algorithm
+ * 2) db_salt->salt depends on host system (e.g. different on 32 and 64-bit)
+ * 3) This salt internally has tunable costs and is used to
+ * populate db_salt->cost[FMT_TUNABLE_COSTS].
+ *
+ * Salt is sent to devices in some uniform way in CMP_CONFIG packet:
+ * - first it goes binary salt in network byte order
+ * - then it sends 4-byte tunable cost(s) if any
+ * - then it sends 2 bytes - number of hashes
+ * - then partial hashes ("binaries") sorted in ascending order.
+ * (It accepts up to 512 32-bit hashes, order doesn't matter).
+ */
+
+static int crypt_all(int *pcount, struct db_salt *salt)
+{
+	int result;
+	unsigned char salt_buf[18]; // salt to send to device
+
+	int salt_len = strnlen(salt->salt, 8);
+
+	// 1 byte unused, 1 byte salt_len, 16 bytes salt in network byte order
+	salt_buf[0] = 0;
+	salt_buf[1] = salt_len;
+	memcpy(salt_buf + 2, salt->salt, salt_len);
+
+	cmp_config_new(salt, salt_buf, 18);
+
+	result = device_format_crypt_all(pcount, salt);
+	return result;
+}
+
+
+static void *get_salt(char *ciphertext) {
+	return MD5_std_get_salt(ciphertext);
+}
+
+static void *get_binary(char *ciphertext) {
+	return MD5_std_get_binary(ciphertext);
+}
+
+
+extern struct task_list *task_list;
+
+static int cmp_exact(char *source, int index)
+{
+	struct task_result *result = task_result_by_index(task_list, index);
+/*
+	fprintf(stderr,"cmp_exact index %d, key '%s'\n",index,result->key);
+	int i;
+	for (i=0; i < 16; i++)
+		fprintf(stderr,"%02x ", result->binary[i]);
+	fprintf(stderr,"\n");
+	for (i=0; i < 16; i++)
+		fprintf(stderr,"%02x ", (((char *)get_binary(source)) + i)[0] & 0xFF);
+	fprintf(stderr,"\n");
+*/
+	return !memcmp(result->binary, get_binary(source), 16);
+}
+
+
+static unsigned int iteration_count(void *salt)
+{
+	return 1000;
+}
+
+
+struct fmt_main fmt_ztex_md5crypt = {
+	{
+		FORMAT_LABEL,
+		FORMAT_NAME,
+		ALGORITHM_NAME,
+		BENCHMARK_COMMENT,
+		BENCHMARK_LENGTH,
+		0,
+		PLAINTEXT_LENGTH,
+		BINARY_SIZE,
+		BINARY_ALIGN,
+		SALT_SIZE,
+		SALT_ALIGN,
+		1, // set by device_format_reset()
+		1,
+		FMT_CASE | FMT_8_BIT | FMT_MASK,
+		{
+			"iteration count",
```

Comment:
We shouldn't report any tunable costs for md5crypt.

Reply:
I can do it and I'm afraid that would be time consuming; i've added that (I'd say "a hack") to simplify development.
The problem is: while a program for computing units' CPUs is selected at runtime (md5crypt or phpass), input to computing units (from the host via cmp_config and arbiter_tx) is common for both programs, including iteration count.

I can do one of the following:
- introduce different salt types on the FPGA side in common bitstream: while phpass salt would pass-by the count from the host, md5crypt salt would add-up 1,000 in 4 bytes during processing;
- don't modify bitstream, send 1,000 as iteration count from the host and somehow keep it invisible for a user, exclude ```iteration_count``` from ```struct fmt_main```.

Please confirm if this modification is required.

For the future development, I have thoughts on some "I/O processor" that would handle input/output to/from the rest of the FPGA, perform basic preparations, simplify further jtr-fpga development by a replacement of custom I/O curcuits with CPU programs.

Reply:
I'm surprised you find fixing this time-consuming. Of course, I meant the second option: "don't modify bitstream, send 1,000 as iteration count from the host and somehow keep it invisible for a user, exclude iteration_count from struct fmt_main." I think it should be do-able in under an hour of effort, no?

Reply:
This hack seems to work:

```diff
diff --git a/src/ztex/pkt_comm/cmp_config.c b/src/ztex/pkt_comm/cmp_config.c
index dda95ab..be2c1e9 100644
--- a/src/ztex/pkt_comm/cmp_config.c
+++ b/src/ztex/pkt_comm/cmp_config.c
@@ -102,7 +102,8 @@ void cmp_config_new(struct db_salt *salt, void *salt_ptr, int salt_len)
 
        int cost_num;
        for (cost_num = 0; cost_num < FMT_TUNABLE_COSTS; cost_num ++) {
-               if (jtr_fmt_params->tunable_cost_name[cost_num])
+               if (jtr_fmt_params->tunable_cost_name[cost_num] ||
+                   (cost_num == 0 && salt_len == 18) /* md5crypt hack */)
                        cmp_config.tunable_costs[cost_num] = salt->cost[cost_num];
                else
                        cmp_config.tunable_costs[cost_num] = 0;
diff --git a/src/ztex_md5crypt.c b/src/ztex_md5crypt.c
index 53a0017..ce806cd 100644
--- a/src/ztex_md5crypt.c
+++ b/src/ztex_md5crypt.c
@@ -140,6 +140,8 @@ static int crypt_all(int *pcount, struct db_salt *salt)
        salt_buf[1] = salt_len;
        memcpy(salt_buf + 2, salt->salt, salt_len);
 
+       salt->cost[0] = 1000;
+
        cmp_config_new(salt, salt_buf, 18);
 
        result = device_format_crypt_all(pcount, salt);
@@ -174,13 +176,6 @@ static int cmp_exact(char *source, int index)
        return !memcmp(result->binary, get_binary(source), 16);
 }
 
-
-static unsigned int iteration_count(void *salt)
-{
-       return 1000;
-}
-
-
 struct fmt_main fmt_ztex_md5crypt = {
        {
                FORMAT_LABEL,
@@ -197,9 +192,7 @@ struct fmt_main fmt_ztex_md5crypt = {
                1, // set by device_format_reset()
                1,
                FMT_CASE | FMT_8_BIT | FMT_MASK,
-               {
-                       "iteration count",
-               },
+               { NULL },
                {
                        FORMAT_TAG
                },
@@ -213,9 +206,7 @@ struct fmt_main fmt_ztex_md5crypt = {
                fmt_default_split,
                get_binary,
                get_salt,
-               {
-                       iteration_count,
-               },
+               { NULL },
                fmt_default_source,
                {
                        fmt_default_binary_hash_0,
```

Reply:
Maybe it'd be less of a hack to start supporting "nameless" tunable costs, which would thus be invisible to the user, but would work for this code anyway. Perhaps this will amount to removal of the name (set it to NULL like above), but not the function, and checking `tunable_cost_value` (function pointers) instead of `tunable_cost_name` in `cmp_config_new`.

Reply:
I'm kind of surprised that you look into the code so in-depth.
I've got other solution that would not modify shared code for ztex formats,
limited to ztex_md5crypt.c source.